File: java/word-count/src/main/java/WordCount.java
import java.text.BreakIterator;
import java.util.HashMap;
import java.util.Map;

class WordCount {

    public Map<String, Integer> phrase(final String sentence) {
        final Map<String, Integer> wordCountMap = new HashMap<>();
        final var wordBreakIterator = BreakIterator.getWordInstance();
        wordBreakIterator.setText(sentence);

        for (int wordStartIndex = wordBreakIterator.first(), wordEndIndex = wordBreakIterator.next();
             wordEndIndex != BreakIterator.DONE;
             wordStartIndex = wordEndIndex, wordEndIndex = wordBreakIterator.next()) {
            final String word = sentence.substring(wordStartIndex, wordEndIndex).toLowerCase();

            if (isWord(word)) {
                wordCountMap.merge(word, 1, Math::addExact);
            }
        }

        return wordCountMap;
    }

    private boolean isWord(final String word) {
        if (word.length() == 1) {
            return Character.isLetterOrDigit(word.charAt(0));
        }

        return word.trim().length() > 0;
    }
}

File: kotlin/word-count/src/main/kotlin/WordCount.kt
private val PUNCTUATION = setOf('!', '&', '@', '$', '%', '^', ':', '.')
private val DELIMITERS = charArrayOf(' ', ',')

private fun String.sanitize() = this.trim().filterNot { c -> c in PUNCTUATION }

private fun String.isEnclosedWithParenthesis() = this.startsWith("'") && this.endsWith("'")

private fun String.removeParenthesis() = this.substring(1, this.length - 1)

private fun putWordIntoMap(map: MutableMap<String, Int>, word: String): MutableMap<String, Int> {
    map.merge(
        if (word.isEnclosedWithParenthesis()) word.removeParenthesis() else word,
        1,
        Math::addExact
    )

    return map
}

object WordCount {
    fun phrase(phrase: String) = phrase
        .toLowerCase()
        .split(*DELIMITERS)
        .map(String::sanitize)
        .filter(String::isNotEmpty)
        .fold(mutableMapOf(), ::putWordIntoMap)
}
